### PR TITLE
RavenDB-19668: use different schemas for document storage

### PR DESCRIPTION
**Pull request by garayx**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-19668

### Additional description

Use different documents (and document additions) schemas for sharded and regular databases.

### Type of change

- Optimization

### How risky is the change?

- Low 

### Backward compatibility

- Not relevant

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing 

- It has been verified by manual testing

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- It requires further work in the Studio

### CODE REVIEW

**Comment by karmeli87** [comment]
We probably want to have this PR regardless, but I'm not expecting that sharding will have a 50% performance penalty on the storage level.

**Comment by ayende** [comment]
I actually would rather *not* have this PR.
The issue is that when we modify the schema, we'll need to apply that across the whole matrix.
The performance cost is compelling, but let's first figure out *why* this is the case?

File: src/Raven.Server/Documents/Sharding/ShardedDocumentsStorage.cs [hunk 1]
```diff
@@ -54,7 +54,24 @@ protected override DocumentPutAction CreateDocumentPutAction()
...
+    protected override void SetDocumentsStorageSchemas()
+    {
+        DocsSchema = Schemas.Documents.ShardingDocsSchemaBase;
+        TombstonesSchema = Schemas.Tombstones.ShardingTombstonesSchema;
+        CompressedDocsSchema = Schemas.Documents.ShardingCompressedDocsSchemaBase;
+    }
+
+    protected override long ReadLastAttachmentsEtagInternal(Transaction tx)
+    {
+        var table = tx.OpenTable(ShardingAttachmentsSchemaBase, AttachmentsMetadataSlice);
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Sharding/ShardedDocumentsStorage.cs]: sth is off here, you are reading inside DocumentsStorage the AttachmensEtag? shouldn't we use the AttachmentStorage?

Comment (ppekrol) [reply src/Raven.Server/Documents/Sharding/ShardedDocumentsStorage.cs]: I see you removed that from the AttachmentsStorage, not sure why?

**Review by ppekrol — APPROVED** [review]

File: src/Raven.Server/Documents/AttachmentsStorage.cs [hunk 2]
```diff
@@ -30,38 +31,35 @@ namespace Raven.Server.Documents
...
-        internal static readonly TableSchema AttachmentsSchema = Schemas.Attachments.Current;
+        internal readonly TableSchema AttachmentsSchema;
 
         private readonly DocumentDatabase _documentDatabase;
         private readonly DocumentsStorage _documentsStorage;
 
         public AttachmentsStorage(DocumentDatabase documentDatabase, Transaction tx)
         {
             _documentDatabase = documentDatabase;
+            if (_documentDatabase is ShardedDocumentDatabase)
```
Comment (karmeli87) [inline src/Raven.Server/Documents/AttachmentsStorage.cs:44]: maybe it is time to split the classes completely?
Right now we have `AttachmentsStorage.cs`  & `AttachmentsStorage.Sharding.cs` which are partial of the same class.
@ppekrol thought?

Comment (garayx) [reply src/Raven.Server/Documents/AttachmentsStorage.cs:44]: @ppekrol will do that in the future

File: src/Raven.Server/Documents/Sharding/ShardedDocumentsStorage.cs (hunk 1 above)
Comment (garayx) [reply src/Raven.Server/Documents/Sharding/ShardedDocumentsStorage.cs]: we actually read all storages (counter, ts, etc) last etags in `DocumentsStorage`,

Comment (garayx) [reply src/Raven.Server/Documents/Sharding/ShardedDocumentsStorage.cs]: done

File: src/Raven.Server/Documents/AttachmentsStorage.cs (hunk 2 above)
Comment (karmeli87) [reply src/Raven.Server/Documents/AttachmentsStorage.cs:44]: https://issues.hibernatingrhinos.com/issue/RavenDB-19722/Sharding-Schema-tasks

**Pull request merged**